src/Algebra/Graph/AdjacencyIntMap/Algorithm.hs

```diff
@@ -40,65 +40,41 @@ import Data.Tree
--- 'forest' (bfsForest ('circuit' [1..5] + 'circuit' [5,4..1])) == 'path' [1,2,3] + 'path' [1,5,4]
--- @
-bfsForest :: AdjacencyIntMap -> Forest Int
-bfsForest g = bfsForestFrom' (vertexList g) g
-
--- | Like 'bfsForest', but the traversal is seeded by a list of
---   vertices. Vertices not in the graph are ignored.
+--   to their 'Ord' instance. The search is seeded by a list of
+--   argument vertices that will be the roots of the resulting
+--   forest. Argument vertices not in the graph are ignored.
 --
 --   Let /L/ be the number of seed vertices. Complexity:
 --   /O((L+m)*min(n,W))/ time and /O(n)/ space.
 --
 -- @
--- 'forest' (bfsForestFrom [1,2] $ 'edge' 1 2)      == 'vertices' [1,2]
--- 'forest' (bfsForestFrom [2]   $ 'edge' 1 2)      == 'vertex' 2
--- 'forest' (bfsForestFrom [3]   $ 'edge' 1 2)      == 'empty'
--- 'forest' (bfsForestFrom [2,1] $ 'edge' 1 2)      == 'vertices' [1,2]
--- 'isSubgraphOf' ('forest' $ bfsForestFrom vs x) x == True
--- bfsForestFrom ('vertexList' x) x               == 'bfsForest' x
--- bfsForestFrom vs ('vertices' vs)               == 'map' (\v -> Node v []) ('nub' vs)
--- bfsForestFrom [] x                           == []
--- bfsForestFrom [1,4] (3 * (1 + 4) * (1 + 5))  == [ Node { rootLabel = 1
+-- 'forest' (bfsForest [1,2] $ 'edge' 1 2)      == 'vertices' [1,2]
+-- 'forest' (bfsForest [2]   $ 'edge' 1 2)      == 'vertex' 2
+-- 'forest' (bfsForest [3]   $ 'edge' 1 2)      == 'empty'
+-- 'forest' (bfsForest [2,1] $ 'edge' 1 2)      == 'vertices' [1,2]
+-- 'isSubgraphOf' ('forest' $ bfsForest vs x) x == True
+-- bfsForest vs ('vertices' vs)               == 'map' (\v -> Node v []) ('nub' vs)
+-- bfsForest [] x                           == []
+-- bfsForest [1,4] (3 * (1 + 4) * (1 + 5))  == [ Node { rootLabel = 1
 --                                                        , subForest = [ Node { rootLabel = 5
 --                                                                             , subForest = [] }]}
 --                                                 , Node { rootLabel = 4
 --                                                        , subForest = [] }]
--- 'forest' (bfsForestFrom [3] ('circuit' [1..5] + 'circuit' [5,4..1])) == 'path' [3,2,1] + 'path' [3,4,5]
+-- 'forest' (bfsForest [3] ('circuit' [1..5] + 'circuit' [5,4..1])) == 'path' [3,2,1] + 'path' [3,4,5]
 -- 
 -- @
-bfsForestFrom :: [Int] -> AdjacencyIntMap -> Forest Int
-bfsForestFrom vs g = bfsForestFrom' [ v | v <- vs, hasVertex v g ] g
+bfsForest :: [Int] -> AdjacencyIntMap -> Forest Int
+bfsForest vs g = bfsForest' [ v | v <- vs, hasVertex v g ] g
 
-bfsForestFrom' :: [Int] -> AdjacencyIntMap -> Forest Int
-bfsForestFrom' vs g = evalState (explore vs) IntSet.empty where
-  explore (v:vs) = discovered v >>= \case
-    True -> (:) <$> unfoldTreeM_BF walk v <*> explore vs
-    False -> explore vs
-  explore [] = return []
+bfsForest' :: [Int] -> AdjacencyIntMap -> Forest Int
+bfsForest' vs g = evalState (explore vs) IntSet.empty where
```

Comment:
I'd probably just do vertex filtering right here to avoid the indirection, i.e.

```haskell
bfsForest vs g = evalState (explore [ v | v <- vs, hasVertex v g ]) IntSet.empty where
```

Reply:
Makes sense!